---
title: "NGS_Final"
author: "Adia Redd"
date: "2025-05-01"
output:
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

This report will describe the results of differential gene expression (DGE) performed on RNA sequencing data that was generated in 2018 by Jiao et al., who treated the MDA-MB-231 breast cancer cell line with a control-siRNA and a treatment-siRNA. The treatment-siRNA, si-NRDE2-2, is meant to block the splicing and translation of the NRDE2 gene, which is necessary for successful cell proliferation. The experiment consists of three biological replicates of both control-siRNA and si-NRDE2-2 treated cells and will be referred to as 'control' and 'RNAi' throughout the remainder of this report. DGE is an important tool for researchers like Jiao et al. to identify statistically significant differences in the mRNA content of cells, which can inform of their transcriptional state. The degree to which genes are expressed is directly correlated to the behavior of cells and in the case of cancers, can help identify where in the genome malignancies and therapy targets may lie. Therefore, performing such analysis is critical to understanding key differences in healthy and diseased cells. DESeq2 is a widely used R package for conducting DGE analysis and will be implemented for this report.

# Materials and Methods

To perform DGE on the mRNA data generated by Jiao et al., the six fastq files for the three control and three RNAi replicates were programmatically downloaded onto my personal scratch hosted on NYU's HPC. The files were retrieved from the Sequence Read Archive hosted by the NCBI using the sbatch script listed under appendix item A. Successful download of the files was confirmed using the command line listed under appendix item B, which reports an exit status of 0 for each fastq file if the command was executed without error. Files ending in numbers between 0-2 store RNA sequencing data for the control replicates, while files ending in numbers between 3-5 store that of the RNAi replicates.

Prior to performing DGE analysis, the mRNA sequences need to be counted. The gold-standard method for performing this task with the intent of analyzing the data using DESeq2 is gathering gene-level abundance estimates using Salmon. Salmon is used to perform pseudoalignment of the cDNA in the fastq files to the human genome so that those estimates can be generated in a quick and storage-efficient manner. Salmon was run using the pipeline stored on nf-core for RNAseq and the sbatch script that was executed is listed under appendix item C1 and the sample sheet used as input for the RNAseq pipeline is provided as well under appendix item C2. Because this experiment contained single-ended reads, the sample sheet is blank in the fastq_2 column, as there is no second fastq file for any of the samples. Strandedness was set to auto, as this feature is not reported in the publication. Nextflow version 24.10.4 and nf-core/rnaseq version 3.18.0 were implemented in this script. The reference genome FASTA and GTF file were obtained using the recommended command line provided on nf-core and is listed under appendix item C3. Additional arguments were added to the rnaseq command line for Salmon quantification to account for any GC biases in the RNA reads. A config file was specified at the command line and is the one provided by the instructor in a previous run of this pipeline. However, upon further investigation it appears this may have been unnecessary because the HPC account had already been defined in the sbatch directives and it is not apparent if the other specifications were pertinent to execute the pipeline. The params file was also borrowed from the instructor, where it was defined that Salmon would be used as the pseudoaligner. The save_reference flag was set to true, but this was not necessary either because Salmon was used and not STAR, which is defined in the params file as well. Nevertheless, the pipeline was successfully run, as confirmed by the command under appendix item B.

The output of the nf-core/rnaseq pipeline included, but was not limited to, a MultiQC report, an execution report, and quant.sf files for each replicate. The MultiQC and execution reports are included separately, and the software versions found in the MultiQC report can be found in appendix item D. As a part of the pipeline, the reads were trimmed using the default trim tool, TrimGalore!.

The R package Tximport was used to summarize the transcript-level counts into gene-level estimations for use in DESeq2, following the outline for best practices established by Soneson et al. To do this and subsequent analysis, RStudio version 2024.12.1+563 running on R version 4.4.1 was used. The packages necessary for analysis can be found in appendix item E. Initialization of variables to represent the metadata of the samples, like ID and respective condition, can be viewed in appendix item F. This is also where the quantification files and tx2gene file generated by Salmon were read into the environment and used to create the tximport object, txi, that will be used as the data set for DESeq2. As shown in appendix item G, DESeq2 was run on this object, with the columns being populated with information stored in the metadata data frame. The experimental design is defined as '\~condition’ because this the feature that separates the sets of replicates. Genes with low expression, defined as less than 10 reads between all six samples, were removed and DESeq was called on this data set.

To perform Principal Component Analysis, the read counts were first normalized using the rlog transformation, then the plotPCA function was called on the transformed data (appendix item H). Next, a results object was generated with the contrast defined as treated/control for the logfoldchange (LFC) calculations and an FDR/alpha of 0.05 (appendix item I). The LFCs were shrunk as well to improve investigation and visualization of the data using the apeGLM model (appendix item J). Several other features of the data were extracted (appendix items K-M), including a plot of dispersion estimates (appendix item G2).

# Results and Discussion

The MultiQC report states that all 6 samples show concerning trends for one metric and completely fail two metrics. The metric with a warning is the sequence length distribution, which is issued when the reads in a sample are not of the same length. This metric was passed prior to trimming, but after trimming all samples have reads that are between 70 and 74 bases long. This is to be expected, and any concern should be assuaged by the fact that the variation in length coincides with the length of trimmed adapter sequences reported in the Cutadapt portion of the report. The samples failed the Per Sequence Base content metric, as all of the samples have wildly varying base calls at numerous positions at the beginning and end of the reads. However, this does not disqualify the data set because, according to the FastQC documentation, this behavior is expected of RNA-seq libraries. This is usually caused by the addition and removal of primers necessary for cDNA transcription and should not affect downstream analysis. The samples also failed the Sequence Duplication Levels metric. The graph displays that about 20-25 percent of the reads, depending upon the sample, occur once but another \~45% of the library is experiencing relative duplication levels between about 10 and 1000. Failure of this metric is not grounds for disqualification either based on the input of Istvan Albert, maintainer of the Biostars forum and author of the Biostars Handbook. It was stated that this trend can be attributed to the fact that upregulation of a gene can cause extremely high levels of dominant transcripts, and this enrichment may appear as sequence duplication. Under appendix item K is the code used to generate the mapping rates for each sample, using counts values that were normalized using the estimate size factors function for a more accurate evaluation. Each sample contributes equally to the total number of reads, with the percent of reads from each being about 16.6%. With the passing of all other MultiQC metrics and similar mapping rates, this data set is suitable for further statistical analysis.

The DESeq function in this package is meant to form a linear model which will be used to test the differences in gene expression between samples for statistical significance. To confirm that the model is suitable for the data, a plot of dispersion estimates was generated. Jiao et al.’s RNA-seq data produces a prototypical plot, according to the DESeq2 documentation. The theoretical shrinkage of dispersion estimates was successful, with many genes being pulled closer to the model and outliers being left alone. Convergence of the model once again indicates suitability for further analysis.

Plotting of the principal component analysis results reveals that the control and treated samples from distinct groups within the first component and minimally overlap within the second component. The expectation would be that SRR7819995 groups higher on the y-axis with the rest of the treated samples, but it is significantly lower. Because PCA is just an abstraction of the data, it is difficult to tell what is causing this deviation. This is especially true because the sample correlation heat map from the MultiQC report (appendix item H2), does not indicate there is a stark difference between this treated replicate and the others. Regardless, deeper investigation of sample SRR7819995 is warranted, perhaps from a proteomics perspective.

The DESeq2 results function was called on the dds object, then transformed into a data frame, allowing for the generation of a histogram of the p-values. The p-values indicate the statistical significance of the LFC in expression of a particular gene between the treated and control groups without corrections for multiple testing. It is observed that there are a large number differentially expressed genes, with the unimodal distribution skewed to the left of the dashed vertical line that indicates the threshold of the chosen alpha value. Plotting of the data did generate a warning that 24 of the rows of the results data frame were excluded from the plot, which was confirmed to be because a p-value had not been calculated for these values. Per the DESeq2 documentation this is because they were considered to be extreme outliers. The skewed distribution does preliminarily demonstrate there is indeed a difference in expression of at least some genes between the control and treated samples.

LFC shrinkage was performed (appendix item J) using the apeGLM method. For transparency, a warning was issued that the apeGLM model was not able to complete the line search routine or sufficiently reduce function value. After reading some of Michael Love's suggestions on the matter and applying the suggestions made to resolve this, it appears the model is still having the same issue. It may not be a good fit for this data set, so shrinkage using the ashr model was tested. The resulting MA plots from each method were compared to determine which would be utilized for analysis. Based on the observations that the models look practically the same and the fact that only a warning was issued, as opposed to an error, the apeGLM model was used. The plot shows several statistically significant genes without the noise of their lower expressing counterparts.

To find the top 10 differentially expressed genes, the p-adjusted column of the results data frame was organized in ascending order and the top 10 were isolated (appendix item L1). The use of p-adjusted values ensures that any genes that did not pass the independent filtering don’t contribute to the significance calculation. The names of the genes were pulled from the tx2gene table and are listed from most differentially to less differentially expressed: BANF1, H3-3A, PTPN1, MET, CALU, JAG1, VAPM7, STMN1, BOD1, and NRAS. I thought it would be interesting to see what purpose these genes serve to provide context as to why they’ve been attributed such significance despite lower LFC values. To this end, a Gene Ontology (GO) term enrichment was performed (appendix item L2) using the default p-value adjustment method (Benjamini-Hochberg), and the top 10 terms extracted. Almost all of the terms are related to negative regulation of signaling pathways and skin development. This is sensible, as the experiments aimed to silence a mRNA, which are often used for transcriptional regulation. One feature that can be linked directly to the experiment performed in the paper is statistically significant changes in expression of molecules that control actin filament bundle assembly, causing negative regulation of this process. NRDE2 plays a role in the expression proteins needed to recruit key centrosomal proteins required to progress through mitosis. This could possibly include those needed to form actin filaments used for cellular reshaping, a process that is characteristic of mitosis.

Finally, of the genes whose adjusted p-value were equal to 0.05 or less, those who experienced the greatest log fold change in expression were isolated. Upon manual inspection of the head and tail of the data frame (appendix item M), there are four genes that experienced a much greater change than the others by 3-4 orders of magnitude. Those genes are: TGFB2-OT1, ENSG00000196826, ZNF658B, and RPS26P58. Their respective LFC values are 10.988413, 9.194353, -8.140350, and -10.903824. Searches of these gene IDs in the NCBI Gene Database show that TGFB2-OT1 is a long-noncoding RNA (lncRNA); ZNF658B is a pseudogene coding for zinc finger protein; and RPS26P58 is another pseudogene coding for 40S Ribosomal Protein S26. When querying ENSG00000196826 initially, I did not get much information. This is because the names of these proteins were pulled from the tx2gene file that was read into R, and it lists that the gene name and gene ID for ENSG00000196826 are the same. However, upon further investigation, I found that ENSG00000196826 refers to ZNF709, another zinc finger protein, and has a very similar summary to ZNF658B in the gene database. TGFB2-OT1 is associated with the diseases Septic Myocarditis and Retinitis Pigmentosa, while ZNF709’s and ZNF658B’s NBCI gene entry lists several recent predictions of what they should do, likely based on conserved sequences. Zinc finger protein and lncRNA substrates tend to be double-stranded DNA, which is usually observed in a regulatory context. Still there is not much to be found on RPS26P58, whose RefSeq status is inferred while the others have been validated.

# Conclusion

In conclusion, DESeq2 analysis of breast cancer cells treated with a control si-RNA and si-NRDE2-2, meant to silence transcription of NRDE2 gene, exhibit large and/or statistically significant changes in gene expression of the following genes: BANF1, H3-3A, PTPN1, MET, CALU, JAG1, VAPM7, STMN1, BOD1, NRAS, TGFB2-OT1, ENSG00000196826, ZNF658B, and RPS26P58. Almost all gene products appear to be regulatory in nature based on querying of the genes on NCBI and GO term enrichment. This analysis corroborates the findings of the paper from which the data was borrowed, which is that the absence of NRDE2 causes issues with RNA splicing and genomic instability. Such an analysis provides a starting point for further investigation into the relevance of this highly conserved protein in the human genome.

# Appendix

## A

```{bash eval=FALSE}

#!/bin/bash
#
#SBATCH --nodes=1
#SBATCH --tasks-per-node=1
#SBATCH --cpus-per-task=1
#SBATCH --time=5:00:00
#SBATCH --mem=8GB
#SBATCH --job-name=download_fastqs
#SBATCH --account=pr_284_general

module purge

wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR781/000/SRR7819990/SRR7819990.fastq.gz
echo _ESTATUS_ [ wget SRR7819990 ]: $?
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR781/001/SRR7819991/SRR7819991.fastq.gz
echo _ESTATUS_ [ wget SRR7819991 ]: $?
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR781/002/SRR7819992/SRR7819992.fastq.gz
echo _ESTATUS_ [ wget SRR7819992 ]: $?
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR781/003/SRR7819993/SRR7819993.fastq.gz
echo _ESTATUS_ [ wget SRR7819993 ]: $?
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR781/004/SRR7819994/SRR7819994.fastq.gz
echo _ESTATUS_ [ wget SRR7819994 ]: $?
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR781/005/SRR7819995/SRR7819995.fastq.gz
echo _ESTATUS_ [ wget SRR7819995 ]: $?
echo _END_ [ download.slurm ]: $(date)
```

## B

```{bash DownloadConfirmation_command, eval=FALSE}
grep _ESTATUS_ slurm*out
```

## C

### 1)

```{bash SalmonPipeline_SBATCH_Script, eval=FALSE}
#!/bin/bash
#
#SBATCH --nodes=1
#SBATCH --tasks-per-node=1
#SBATCH --cpus-per-task=4
#SBATCH --time=8:00:00
#SBATCH --mem=4GB
#SBATCH --job-name=rnaseq
#SBATCH --mail-type=FAIL
#SBATCH --mail-user=ar9481@nyu.edu
#SBATCH --account=pr_284_general

module purge
module load nextflow/24.10.4

nextflow run nf-core/rnaseq -r 3.18.0 \
        --input samplesheet.csv \
        --outdir results \
        --fasta Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa.gz \
        --gtf Homo_sapiens.GRCh38.113.gtf.gz \
        --extra_salmon_quant_args \"--gcBias\" \
        --save_reference TRUE \
        -profile nyu_hpc \
        --account=pr_284_general \
        -c /scratch/work/courses/BI7653/hw8.2025/rnaseq.config \
        -params-file /scratch/work/courses/BI7653/hw8.2025/rnaseq.json \
        -resume 

echo _ESTATUS_ [ rnaseq ]: $?
echo _END_ [ rnaseq ] : $(date)
```

### 2)

```{r samplesheet}
   # Reading in the samplesheet used for nf-core/rnaseq execution
samplesheet <-read.csv('samplesheet.csv')
samplesheet
```

### 3)

```{bash eval=FALSE}
latest_release=$(curl -s 'http://rest.ensembl.org/info/software?content-type=application/json' | grep -o '"release":[0-9]*' | cut -d: -f2)
wget -L ftp://ftp.ensembl.org/pub/release-${latest_release}/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa.gz
wget -L ftp://ftp.ensembl.org/pub/release-${latest_release}/gtf/homo_sapiens/Homo_sapiens.GRCh38.${latest_release}.gtf.gz
```

## D

```{r}
x <- '
Group	Software	Version
CUSTOM_GETCHROMSIZES	getchromsizes	1.21
CUSTOM_TX2GENE	python	3.10.4
DESEQ2_QC_PSEUDO	bioconductor-deseq2	1.28.0
-\tr-base	4.0.3
FASTQC	fastqc	0.12.1
FQ_LINT	fq	0.12.0 (2024-07-08)
FQ_SUBSAMPLE	fq	0.12.0 (2024-07-08)
GTF2BED	perl	5.26.2
GTF_FILTER	python	3.9.5
GUNZIP_FASTA	gunzip	1.1
GUNZIP_GTF	gunzip	1.1
MAKE_TRANSCRIPTS_FASTA	rsem	1.3.1
-\tstar	2.7.10a
SALMON_INDEX	salmon	1.10.3
SALMON_QUANT	salmon	1.10.3
SE_GENE	bioconductor-summarizedexperiment	1.32.0
TRIMGALORE	cutadapt	4.9
-\ttrimgalore	0.6.10
TXIMETA_TXIMPORT	bioconductor-tximeta	1.20.1
Workflow	Nextflow	24.10.4
-\tnf-core/rnaseq	v3.18.0-gb96a753
'
multiqc_software_versions <- read.table(text = x, sep = '\t', header = TRUE, strip.white = TRUE)
multiqc_software_versions
```

## E

```{r message=FALSE, warning=FALSE}

  # Calling necessary packages 
library(DESeq2)
library(tximport)
library(ggplot2)
library(tidyverse)
library(ggrepel)
library(clusterProfiler)
library(AnnotationDbi)
library(org.Hs.eg.db)
```

## F

```{r}

  # Creation of variables to store metadata 
sample_id <- c('SRR7819990','SRR7819991','SRR7819992','SRR7819993','SRR7819994','SRR7819995')
cond <- c(rep('control',3),rep('treated',3))
quant_files <- file.path('salmon',sample_id,'quant.sf')
names(quant_files) <- sample_id
tx2gene <- read.csv(file.path('salmon','tx2gene.tsv'),sep = '\t')
```

```{r message=FALSE}
  
  # Creating txi object from quant files and tx2gene file generated by the pipeline
txi <- tximport(quant_files, type="salmon", tx2gene=tx2gene)
```

```{r}

  # Creation of metadata dataframe to be used in creation of DESeq data set
metadata_df <- data.frame(sample = factor(sample_id),condition = factor(cond, levels = c('control','treated')) )
row.names(metadata_df) <- sample_id
metadata_df
```

## G

```{r message=FALSE}
  # Creation of DDS from txi obejct and isolation of genes with greater than or equal to 10        reads

dds <- DESeqDataSetFromTximport(txi,
                                   colData = metadata_df,
                                   design = ~ condition)

keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
dds <- DESeq(dds)
```

```{r}
  # Dispersion Estimate Plot
plotDispEsts(dds)
```

## H

### 1)

```{r message=FALSE}
  # Transformation of data set using rlog method
rltrans <- rlog(dds)

  # Plotting of PCA components 
pcaData <- plotPCA(rltrans, intgroup = 'condition', returnData=TRUE)

  # Getting the percent of variance each of the two components capture
percentVar <- round(100 * attr(pcaData, "percentVar"))
cust_pcaplot <- ggplot(pcaData, aes(PC1, PC2, color=condition)) +
                  geom_point(size=3) +
                  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
                  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
                  coord_fixed()
  
  # Adding point labels to the plot
cust_pcaplot+
  geom_label_repel( aes(label = sample_id),
                  box.padding   = 0.35, 
                  point.padding = 0.5)
```

### 2)

![](images/salmon_deseq2_clustering-plot.png)

## I

```{r}
  # Creation of results object
rez <- as.data.frame(results(dds, contrast = c('condition','treated','control'), alpha = 0.05))
  
  # Putting gene IDs in their own column
rez <- rez %>%
        rownames_to_column(var = "GeneID")

  # P-value histogram with vertical axis at alpha value
ggplot(rez, aes(x = pvalue ))+
  geom_histogram(bins = 75)+
  geom_vline(xintercept =  0.05, linetype = 'dashed', color = 'forestgreen')
```

## J

```{r}
  # Shrinking LFCs with ashr and apeGLM methods and creating MA plots for both

resLFC_apeglm <- lfcShrink(dds, coef="condition_treated_vs_control", type="apeglm")
resLFC_ashr <- lfcShrink(dds, contrast =c('condition','treated','control'), type="ashr")

plotMA(resLFC_apeglm, main = 'apeGLM MA Plot')
plotMA(resLFC_ashr, main = 'ashr MA Plot')
```

## K

```{r message=FALSE}
  # Normalization of read counts and storing values in data frame 
dds1 <- estimateSizeFactors(dds)
norm_counts_df <- as.data.frame(counts(dds1, normalized = TRUE ))
normcounts_long <- norm_counts_df %>%
  rownames_to_column("gene") %>%
  pivot_longer(
    cols = -gene, 
    names_to = "sample_id",
    values_to = "counts")
```
```{r}
  # Mapping rate calculation
total <- sum( colSums(norm_counts_df) )
mapping_rates <- as.data.frame( colSums(norm_counts_df)/total )
colnames(mapping_rates)[colnames(mapping_rates) == 'colSums(norm_counts_df)/total'] <- '% of Total Reads'
mapping_rates*100   # converting rates into percentages

```

## L

### 1)

```{r warning=FALSE}
  # Top 10 DGE
top_10_dge <- rez[head(order(rez$padj),10),]
top_10_dge
```
```{r warning=FALSE}
  # Extraction of gene names from ENSEMBL IDs
top10names <- list()
for (gene in top_10_dge$GeneID){
  top10names[gene] <- tx2gene$gene_name[tx2gene$gene_id == gene]
}
print(c('The top 10 differentially expressed genes are',top10names ))
```

### 2)

```{r}
  # GO term enrichment of the top 10 DGE
GO_results <- enrichGO(gene = top_10_dge$GeneID, OrgDb = 'org.Hs.eg.db', keyType = 'ENSEMBL', ont = 'BP')

  # Plot of top 10 enriched terms
plot(barplot(GO_results, showCategory=10, font.size = 8))
```

## M

```{r}
  # Isolation of genes whose adjusted p-value are less than or equal to the alpha defined in
  # results, and ordering them from greatest to least LFC
total_statsig <- rez[!is.na(rez$padj),]
total_statsig <- total_statsig[total_statsig$padj <= 0.05,]
total_statsig <- total_statsig[ order(total_statsig$log2FoldChange, decreasing = TRUE),]

head(total_statsig,10)
tail(total_statsig,10)
```
```{r}
  # Highest and lowest changes in expression
relevantrows <- rbind( head(total_statsig,2), tail(total_statsig,2) )
```
```{r}
  # Extraction of gene names from ENSEMBL IDs
mostchanged <- list()
for (gene in relevantrows$GeneID){
  mostchanged[gene] <- tx2gene$gene_name[tx2gene$gene_id == gene]
}
mostchanged
```
